src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -3,45 +3,418 @@
+                    nonGcStatics: new StaticsBlock() { Size = nonGcStatics[0], LargestAlignment = blockAlignment },
+                    threadGcStatics: new StaticsBlock() { Size = gcStatics[1], LargestAlignment = blockAlignment },
+                    threadNonGcStatics: new StaticsBlock() { Size = nonGcStatics[1], LargestAlignment = blockAlignment },
+                    typeToFieldMap: typeToFieldMap);
+            }
+
+            protected override int GetKeyHashCode(EcmaModule key)
+            {
+                return key.GetHashCode();
+            }
+
+            protected override int GetValueHashCode(ModuleFieldLayout value)
+            {
+                return value.Module.GetHashCode();
+            }
+
+            /// <summary>
+            /// Try to locate the ThreadStatic custom attribute on the type (much like EcmaField.cs does in the method InitializeFieldFlags).
+            /// </summary>
+            /// <param name="fieldDef">Field definition</param>
+            /// <param name="metadataReader">Metadata reader for the module</param>
+            /// <returns>true when the field is marked with the ThreadStatic custom attribute</returns>
+            private static bool IsFieldThreadStatic(in FieldDefinition fieldDef, MetadataReader metadataReader)
+            {
+                foreach (CustomAttributeHandle customAttributeHandle in fieldDef.GetCustomAttributes())
+                {
+                    StringHandle namespaceHandle, nameHandle;
+                    if (metadataReader.GetAttributeNamespaceAndName(customAttributeHandle, out namespaceHandle, out nameHandle) &&
+                        metadataReader.StringComparer.Equals(namespaceHandle, "System") &&
+                        metadataReader.StringComparer.Equals(nameHandle, "ThreadStaticAttribute"))
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            /// <summary>
+            /// Partially decode field signature to obtain CorElementType and optionally the type handle for VALUETYPE fields.
+            /// </summary>
+            /// <param name="fieldDef">Metadata field definition</param>
+            /// <param name="metadataReader">Metadata reader for the active module</param>
+            /// <param name="corElementType">Output element type decoded from the signature</param>
+            /// <param name="valueTypeHandle">Value type handle decoded from the signature</param>
+            private static void GetFieldElementTypeAndValueTypeHandle(
+                in FieldDefinition fieldDef,
+                MetadataReader metadataReader,
+                out CorElementType corElementType,
+                out Handle valueTypeHandle)
+            {
+                ImmutableArray<byte> signature = metadataReader.GetBlobContent(fieldDef.Signature);
```

Comment:
S.R.Metadata has a `BlobReader` type. We use it in the implementation of the ECMA signature parsing in the type system too. Using that would avoid having to hand roll `ReadCompressedUInt`, `ReadCompressedToken`, or manipulating the array index.